File: scripts/generate-tmc-client-sheet.py
#!/usr/bin/env python3
"""Write TMC client review sheet as .xlsx (requires openpyxl)."""
import json
import sys
from pathlib import Path

try:
    from openpyxl import Workbook
    from openpyxl.styles import Font, PatternFill, Alignment
    from openpyxl.utils import get_column_letter
except ImportError:
    print("openpyxl not installed. Run: pip3 install openpyxl", file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parents[1]
PRODUCTS_PATH = ROOT / "public/data/products.json"
MANIFEST_PATH = ROOT / "public/data/tmc-import-manifest.json"
OUT_PATH = ROOT / "docs/TMC_Ring_Import_Client_Sheet.xlsx"

COLUMNS = [
    ("includeOnSite", "Include on ELYSIUM site? (Yes/No)"),
    ("tmcOriginalName", "TMC Original Name"),
    ("tmcShopifyHandle", "TMC Shopify Handle"),
    ("elysiumTitle", "ELYSIUM Site Name (Title)"),
    ("elysiumSlug", "ELYSIUM URL Slug"),
    ("elysiumUrl", "ELYSIUM Product URL (when live)"),
    ("blurb", "Short Tagline (Blurb)"),
    ("description", "Full Product Description"),
    ("basePriceGBP", "Base Price GBP (from, 1ct lab-grown)"),
    ("featured", "Featured on Homepage? (Yes/No)"),
    ("shape", "Stone Shape"),
    ("styles", "Style Tags (comma-separated)"),
    ("collections", "Collections (comma-separated)"),
    ("metals", "Available Metals (comma-separated)"),
    ("carats", "Carat Options (comma-separated)"),
    ("origins", "Stone Origin Options (Natural / Lab Grown)"),
    ("colours", "Diamond Colour Grades Offered"),
    ("clarities", "Diamond Clarity Grades Offered"),
    ("certificates", "Certification Options (GIA / IGI)"),
    ("engraving", "Engraving Available? (Yes/No)"),
    ("qualityBanner", "Quality Banner Text"),
    ("seoTitle", "SEO Page Title"),
    ("seoDescription", "SEO Meta Description"),
    ("hasYellowImages", "Has Yellow Gold Images? (Yes/No)"),
    ("hasRoseImages", "Has Rose Gold Images? (Yes/No)"),
    ("hasWhitePlatinumImages", "Has White Gold / Platinum Images? (Yes/No)"),
    ("needsPhotography", "Needs New Photography? (Yes/No)"),
    ("clientNotes", "Client Notes / Special Instructions"),
]


def has_metal(product, key):
    return bool(product.get("galleryByMetal", {}).get(key))


def build_rows(products, manifest):
    handle_to_product = {}
    for p in products:
        handle = None
        for img in p.get("images", []):
            if "tmc-import/" in img:
                handle = img.split("tmc-import/")[1].split("/")[0]
                break
        if handle:
            handle_to_product[handle] = p

    site_url = "https://elysium-mvp.vercel.app"
    rows = []
    for entry in manifest["products"]:
        p = handle_to_product.get(entry["handle"])
        if not p:
            continue
        has_yellow = has_metal(p, "18k Yellow Gold")
        has_rose = has_metal(p, "18k Rose Gold")
        has_white = has_metal(p, "18k White Gold") or has_metal(p, "Platinum")
        rows.append({
            "includeOnSite": "",
            "tmcOriginalName": entry["title"],
            "tmcShopifyHandle": entry["handle"],
            "elysiumTitle": p["title"],
            "elysiumSlug": p["slug"],
            "elysiumUrl": f"{site_url}/products/{p['slug']}",
            "blurb": p.get("blurb", ""),
            "description": p.get("description", "").replace("\n", " ").strip(),
            "basePriceGBP": p.get("basePriceGBP", ""),
            "featured": "Yes" if p.get("isFeatured") else "No",
            "shape": p.get("shape", ""),
            "styles": ", ".join(p.get("styles", [])),
            "collections": ", ".join(p.get("collections", [])),
            "metals": ", ".join(m["name"] for m in p.get("metals", [])),
            "carats": ", ".join(c["label"] for c in p.get("carats", [])),
            "origins": ", ".join(o["label"] for o in p.get("origins", [])),
            "colours": ", ".join(c["label"] for c in p.get("colours", [])),
            "clarities": ", ".join(c["label"] for c in p.get("clarities", [])),
            "certificates": ", ".join(c["label"] for c in p.get("certificates", [])),
            "engraving": "Yes" if p.get("engravingMaxChars") else "No",
            "qualityBanner": p.get("qualityBanner", ""),
            "seoTitle": p.get("seoTitle", ""),
            "seoDescription": p.get("seoDescription", ""),
            "hasYellowImages": "Yes" if has_yellow else "No",
            "hasRoseImages": "Yes" if has_rose else "No",
            "hasWhitePlatinumImages": "Yes" if has_white else "No",
            "needsPhotography": "No" if (has_yellow and has_rose and has_white) else "Yes",
            "clientNotes": "",
        })
    return rows


def main():
    products = json.loads(PRODUCTS_PATH.read_text())
    manifest = json.loads(MANIFEST_PATH.read_text())
    rows = build_rows(products, manifest)

    wb = Workbook()
    ws = wb.active
    ws.title = "TMC Ring Import"

    header_fill = PatternFill(start_color="6D3D0D", end_color="6D3D0D", fill_type="solid")
    header_font = Font(color="FFFFFF", bold=True)

    for col_idx, (_, header) in enumerate(COLUMNS, 1):
        cell = ws.cell(row=1, column=col_idx, value=header)
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(wrap_text=True, vertical="top")

    for row_idx, row in enumerate(rows, 2):
        for col_idx, (key, _) in enumerate(COLUMNS, 1):
            ws.cell(row=row_idx, column=col_idx, value=row.get(key, ""))

    for col_idx, (_, header) in enumerate(COLUMNS, 1):
        letter = get_column_letter(col_idx)
        ws.column_dimensions[letter].width = min(48, max(14, len(header) + 2))

    ws.freeze_panes = "A2"
    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    wb.save(OUT_PATH)
    print(f"Wrote {OUT_PATH} ({len(rows)} rings)")


if __name__ == "__main__":
    main()

File: scripts/generate-tmc-client-sheet.mjs
#!/usr/bin/env node
// Generate a client-facing spreadsheet for reviewing TMC ring imports.
// Outputs CSV (opens in Excel) and attempts .xlsx if openpyxl is available.
//
// Usage: node scripts/generate-tmc-client-sheet.mjs

import fs from 'node:fs';
import path from 'node:path';
import { execSync } from 'node:child_process';

const ROOT = process.cwd();
const PRODUCTS_PATH = path.join(ROOT, 'public/data/products.json');
const MANIFEST_PATH = path.join(ROOT, 'public/data/tmc-import-manifest.json');
const OUT_DIR = path.join(ROOT, 'docs');
const CSV_PATH = path.join(OUT_DIR, 'TMC_Ring_Import_Client_Sheet.csv');
const XLSX_PATH = path.join(OUT_DIR, 'TMC_Ring_Import_Client_Sheet.xlsx');

const COLUMNS = [
  { key: 'includeOnSite', header: 'Include on ELYSIUM site? (Yes/No)' },
  { key: 'tmcOriginalName', header: 'TMC Original Name' },
  { key: 'tmcShopifyHandle', header: 'TMC Shopify Handle' },
  { key: 'elysiumTitle', header: 'ELYSIUM Site Name (Title)' },
  { key: 'elysiumSlug', header: 'ELYSIUM URL Slug' },
  { key: 'elysiumUrl', header: 'ELYSIUM Product URL (when live)' },
  { key: 'blurb', header: 'Short Tagline (Blurb)' },
  { key: 'description', header: 'Full Product Description' },
  { key: 'basePriceGBP', header: 'Base Price GBP (from, 1ct lab-grown)' },
  { key: 'featured', header: 'Featured on Homepage? (Yes/No)' },
  { key: 'shape', header: 'Stone Shape' },
  { key: 'styles', header: 'Style Tags (comma-separated)' },
  { key: 'collections', header: 'Collections (comma-separated)' },
  { key: 'metals', header: 'Available Metals (comma-separated)' },
  { key: 'carats', header: 'Carat Options (comma-separated)' },
  { key: 'origins', header: 'Stone Origin Options (Natural / Lab Grown)' },
  { key: 'colours', header: 'Diamond Colour Grades Offered' },
  { key: 'clarities', header: 'Diamond Clarity Grades Offered' },
  { key: 'certificates', header: 'Certification Options (GIA / IGI)' },
  { key: 'engraving', header: 'Engraving Available? (Yes/No)' },
  { key: 'qualityBanner', header: 'Quality Banner Text' },
  { key: 'seoTitle', header: 'SEO Page Title' },
  { key: 'seoDescription', header: 'SEO Meta Description' },
  { key: 'hasYellowImages', header: 'Has Yellow Gold Images? (Yes/No)' },
  { key: 'hasRoseImages', header: 'Has Rose Gold Images? (Yes/No)' },
  { key: 'hasWhitePlatinumImages', header: 'Has White Gold / Platinum Images? (Yes/No)' },
  { key: 'needsPhotography', header: 'Needs New Photography? (Yes/No)' },
  { key: 'clientNotes', header: 'Client Notes / Special Instructions' },
];

function csvEscape(value) {
  const str = String(value ?? '');
  if (/[",\n\r]/.test(str)) return `"${str.replace(/"/g, '""')}"`;
  return str;
}

function hasMetalImages(product, metalKey) {
  return Boolean(product.galleryByMetal?.[metalKey]?.length);
}

function buildRows(products, manifest) {
  const handleToProduct = new Map();
  for (const p of products) {
    const handle = p.images?.[0]?.match(/tmc-import\/([^/]+)/)?.[1];
    if (handle) handleToProduct.set(handle, p);
  }

  return manifest.products.map((entry) => {
    const p = handleToProduct.get(entry.handle);
    if (!p) return null;

    const siteUrl = process.env.NEXT_PUBLIC_SITE_URL || 'https://elysium-mvp.vercel.app';
    const hasYellow = hasMetalImages(p, '18k Yellow Gold');
    const hasRose = hasMetalImages(p, '18k Rose Gold');
    const hasWhite = hasMetalImages(p, '18k White Gold') || hasMetalImages(p, 'Platinum');

    return {
      includeOnSite: '',
      tmcOriginalName: entry.title,
      tmcShopifyHandle: entry.handle,
      elysiumTitle: p.title,
      elysiumSlug: p.slug,
      elysiumUrl: `${siteUrl}/products/${p.slug}`,
      blurb: p.blurb,
      description: p.description?.replace(/\n+/g, ' ').trim(),
      basePriceGBP: p.basePriceGBP,
      featured: p.isFeatured ? 'Yes' : 'No',
      shape: p.shape || '',
      styles: (p.styles || []).join(', '),
      collections: (p.collections || []).join(', '),
      metals: (p.metals || []).map((m) => m.name).join(', '),
      carats: (p.carats || []).map((c) => c.label).join(', '),
      origins: (p.origins || []).map((o) => o.label).join(', '),
      colours: (p.colours || []).map((c) => c.label).join(', '),
      clarities: (p.clarities || []).map((c) => c.label).join(', '),
      certificates: (p.certificates || []).map((c) => c.label).join(', '),
      engraving: p.engravingMaxChars ? 'Yes' : 'No',
      qualityBanner: p.qualityBanner,
      seoTitle: p.seoTitle || '',
      seoDescription: p.seoDescription || '',
      hasYellowImages: hasYellow ? 'Yes' : 'No',
      hasRoseImages: hasRose ? 'Yes' : 'No',
      hasWhitePlatinumImages: hasWhite ? 'Yes' : 'No',
      needsPhotography: hasYellow && hasRose && hasWhite ? 'No' : 'Yes',
      clientNotes: '',
    };
  }).filter(Boolean);
}

function writeCsv(rows) {
  fs.mkdirSync(OUT_DIR, { recursive: true });
  const header = COLUMNS.map((c) => csvEscape(c.header)).join(',');
  const body = rows.map((row) => COLUMNS.map((c) => csvEscape(row[c.key])).join(',')).join('\n');
  // UTF-8 BOM helps Excel open accents correctly
  fs.writeFileSync(CSV_PATH, `\uFEFF${header}\n${body}\n`);
}

function tryWriteXlsx(rows) {
  const pyScript = `
import json, sys
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment
from openpyxl.utils import get_column_letter

rows = json.load(sys.stdin)
columns = json.loads(sys.argv[1])

wb = Workbook()
ws = wb.active
ws.title = "TMC Ring Import"

header_fill = PatternFill(start_color="6D3D0D", end_color="6D3D0D", fill_type="solid")
header_font = Font(color="FFFFFF", bold=True)

for col_idx, col in enumerate(columns, 1):
    cell = ws.cell(row=1, column=col_idx, value=col["header"])
    cell.fill = header_fill
    cell.font = header_font
    cell.alignment = Alignment(wrap_text=True, vertical="top")

for row_idx, row in enumerate(rows, 2):
    for col_idx, col in enumerate(columns, 1):
        ws.cell(row=row_idx, column=col_idx, value=row.get(col["key"], ""))

for col_idx, col in enumerate(columns, 1):
    letter = get_column_letter(col_idx)
    width = min(48, max(14, len(col["header"]) + 2))
    ws.column_dimensions[letter].width = width

ws.freeze_panes = "A2"
wb.save(sys.argv[2])
`;

  try {
    execSync('python3 -c "import openpyxl"', { stdio: 'ignore' });
    execSync(
      `python3 -c ${JSON.stringify(pyScript)} ${JSON.stringify(JSON.stringify(COLUMNS))} ${JSON.stringify(XLSX_PATH)}`,
      { input: JSON.stringify(rows), stdio: ['pipe', 'inherit', 'inherit'] }
    );
    return true;
  } catch {
    return false;
  }
}

function main() {
  const products = JSON.parse(fs.readFileSync(PRODUCTS_PATH, 'utf8'));
  const manifest = JSON.parse(fs.readFileSync(MANIFEST_PATH, 'utf8'));
  const rows = buildRows(products, manifest);

  writeCsv(rows);
  console.log(`Wrote ${CSV_PATH} (${rows.length} rings)`);

  if (tryWriteXlsx(rows)) {
    console.log(`Wrote ${XLSX_PATH}`);
  } else {
    console.log('openpyxl not installed — CSV only (opens in Excel). Run: pip3 install openpyxl');
  }
}

main();
